src/backend/distributed/commands/local_multi_copy.c

```diff
@@ -0,0 +1,203 @@
+/*-------------------------------------------------------------------------
+ *
+ * local_multi_copy.c
+ *    Commands for running a copy locally
+ *
+ * For each local placement, we have a buffer. When we receive a slot
+ * from a copy, the slot will be put to the corresponding buffer based
+ * on the shard id. When the buffer size exceeds the threshold a local
+ * copy will be done. Also If we reach to the end of copy, we will send
+ * the current buffer for local copy.
+ *
+ * The existing logic from multi_copy.c and format are used, therefore
+ * even if user did not do a copy with binary format, it is possible that
+ * we are going to be using binary format internally.
+ *
+ *
+ * Copyright (c) Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+#include "commands/copy.h"
+#include "catalog/namespace.h"
+#include "parser/parse_relation.h"
+#include "utils/lsyscache.h"
+#include "nodes/makefuncs.h"
+#include "safe_lib.h"
+#include <netinet/in.h> /* for htons */
+
+#include "distributed/transmit.h"
+#include "distributed/commands/multi_copy.h"
+#include "distributed/multi_partitioning_utils.h"
+#include "distributed/local_executor.h"
+#include "distributed/local_multi_copy.h"
+#include "distributed/shard_utils.h"
+
+/*
+ * LOCAL_COPY_BUFFER_SIZE is buffer size for local copy.
+ * There will be one buffer for each local placement, therefore
+ * the maximum amount of memory that might be alocated is
+ * LOCAL_COPY_BUFFER_SIZE * #local_placement
```

Comment:
Not exactly true, since this is a threshold (i.e. we flush when buffer->len exceeds this), not maximum size. StringInfo doubles its internal buffer until it is greater than required size, so it can take as much as (2 * LOCAL_COPY_BUFFER_SIZE).

Maybe rename to `LOCAL_COPY_FLUSH_THRESHOLD` to make the naming more accurate.